### ISSUE TITLE
Add support for outputting "metadata" in the render-json command

### ISSUE THREAD

**Issue by brikis98** [gruntwork-io/terragrunt#2196]
Something that would make `render-json` even more powerful for debugging is if the output included not only the fully-resolved values, but also other metadata, such as "breadcrumbs" that tell you where that value was found. This likely means changing the JSON output format, so we'd probably only want this with some flag like `--with-metadata`. 

For example, let's say this is the original `render-json` output (truncated to keep the example simple):

```json
{
  "generate": {
    "provider": {
      "comment_prefix": "# ",
      "contents": "# This is just a test",
      "disable_signature": false,
      "if_exists": "overwrite",
      "path": "provider.tf"
    }
  },
  "inputs": {
    "foo": "bar",
    "baz": "blah"
  },
  "terraform": {
    "source": "./foo"
  }
}
```

Here's what `render-json --with-metadata` might look like:

```json
{
  "generate": {
    "provider": {
      "value": {
        "comment_prefix": "# ",
        "contents": "# This is just a test",
        "disable_signature": false,
        "if_exists": "overwrite",
        "path": "provider.tf",
      },
      "metadata": {
        "found_in_file": "/live/dev/vpc/terragrunt.hcl"
      }
    }
  },
  "inputs": {
    "foo": {
      "value": "bar",
      "metadata": {
        "found_in_file": "/live/dev/vpc/terragrunt.hcl"
      }
    "baz":  {
      "value": "blah",
      "metadata": {
        "found_in_file": "/live/terragrunt.hcl"
      }
  },
  "terraform": {
    "source": {
      "value": "./foo",
      "metadata": {
        "found_in_file": "/live/_envcommon/vpc.hcl"
      }
    }
  }
}
```

In other words, every value because an object with the keys: 

* `value`: The original value goes here.
* `metadata`: Extra metadata goes in this object. Initially, the most useful metadata would be `found_in_file`, which gives you the file path where this value came from: you can see in the examples above, that may be a "child" config (`/live/dev/vpc/terragrunt.hcl`) or a parent config pulled in via `include` (`/live/terragrunt.hcl`, `/live/_envcommon/vpc.hcl`). In the future, we may also want to add metadata such as the name of the `include` that was used to pull in that file, the type of merging done, etc.

**Comment by brikis98** [comment]
@denis256 Do you have any sense of the complexity of implementing this `render-json --with-metadata` functionality?

**Comment by yorinasub17** [comment]
Also, it would be useful for `locals` metadata to indicate what locals are available in the parent config (includes). Even though they may not be available for use, it may be useful for deciding if one should use `expose = true`.

**Comment by denis256** [comment]
Hi,
to add additional fields on `render-json` I thought about:
 * adding of "metadata" map field to [`TerragruntConfig` ](https://github.com/gruntwork-io/terragrunt/blob/master/config/config.go#L33) to store metadata for all fields
 * add code to set metadata during execution
 * during `render-json --with-metadata` fetch metadata for respective field from map and include in output

This approach looks more feasible than changing each `TerragruntConfig` field to have "value" and "metadata"

Implementation shouldn't take more than a week

**Comment by brikis98** [comment]
Roger. Do you have time to start on this now by any chance? Along with the related `render-json` changes in https://github.com/gruntwork-io/terragrunt/issues/2198?

**Comment by denis256** [comment]
Yes, started to work on it
Drafted a [PR](https://github.com/gruntwork-io/terragrunt/pull/2199) which include metadata for locals

Before changes `render-json` :
```
  "locals": {
    "aws_profile": "prod",
    "aws_region": "us-east-1",
    "aws_tags": {
      "test": "test"
    }
  },
```
with changes from PR:
```
  "locals": {
    "aws_profile": {
      "metadata": {
        "found_in_file": "/projects/gruntwork/terragrunt-tests/locals-tf/terragrunt.hcl"
      },
      "value": "prod"
    },
    "aws_region": {
      "metadata": {
        "found_in_file": "/projects/gruntwork/terragrunt-tests/locals-tf/terragrunt.hcl"
      },
      "value": "us-east-1"
    },
    "aws_tags": {
      "metadata": {
        "found_in_file": "/projects/gruntwork/terragrunt-tests/locals-tf/terragrunt.hcl"
      },
      "value": {
        "test": "test"
      }
    }
  },
```

**Comment by brikis98** [comment]
Haha, you are fast. Thank you, that's great 👍 😁

**Comment by denis256** [comment]
Released in https://github.com/gruntwork-io/terragrunt/releases/tag/v0.38.6
